java/src/main/java/com/google/appengine/tools/pipeline/impl/model/JobRecord.java

```diff
@@ -348,42 +343,21 @@ private JobRecord(Key rootJobKey, Key thisKey, Key generatorJobKey, String graph
     if (parentQueueSettings != null) {
       queueSettings.merge(parentQueueSettings);
     }
-    if (queueSettings.getOnBackend() == null) {
-      String module = queueSettings.getOnModule();
-      if (module == null) {
-        String currentBackend = getCurrentBackend();
-        if (currentBackend != null) {
-          queueSettings.setOnBackend(currentBackend);
-        } else {
-          ModulesService modulesService = ModulesServiceFactory.getModulesService();
-          queueSettings.setOnModule(modulesService.getCurrentModule());
-          queueSettings.setModuleVersion(modulesService.getCurrentVersion());
-        }
+    String service = queueSettings.getOnService();
+    ModulesService modulesService = ModulesServiceFactory.getModulesService();
+    if (service == null) {
+      //take current
+      queueSettings.setOnService(modulesService.getCurrentModule());
+      queueSettings.setOnServiceVersion(modulesService.getCurrentVersion());
+    } else {
+      if (service.equals(modulesService.getCurrentModule())) {
+        queueSettings.setOnServiceVersion(modulesService.getCurrentVersion());
```

Comment:
and you need to set the service too, right?
`queueSettings.setOnService(service)`

Personally, I found this code a bit confusing, I'd do:
```java
String service = queueSettings.getOnService();
if (service == null || service.equals(modulesService.getCurrentModule())) {
   queueSettings.setOnService(modulesService.getCurrentModule());
   queueSettings.setOnServiceVersion(modulesService.getCurrentVersion());
} else {
   queueSettings.setOnService(service);
   queueSettings.setOnServiceVersion(modulesService.getDefaultVersion(service));
}
```

Reply:
yeah, it is confusing. I'll try to clean up a little with comments. In effect what this is doing is just setting defaults.  with caveat that you're only allowed to set a specific `serviceVersion` if you explicitly set a `service`